Sorting algos (#9)
Ada-Developers-Academy/core-unit-2: sorting-algos/insertion-sort.md

1. The source list
1. A list into which sorted items are inserted

One entry is removed at a time and then each is inserted into a sorted part (initially empty).

To save memory, most implementations use an in-place sort that works by moving the current item past the already sorted items and repeatedly swapping it with the preceding item until it is in place.

### Detailed Explanation

![Insertion Sort Example](../assets/sorting-algos/insertion-sort.png)

Consider this more detailed explanation of insertion sort:

- Loop through the entire list
- Find the first element of the unsorted list. This is the element that needs to be inserted, called `to_insert`
- Find the correct index to insert the value `to_insert`
  - Loop through the sorted list
  - Compare the values between the item we're inspecting, and `to_insert`
  - Swap if needed
- Increase the indices, so that the sorted list grows, and the unsorted list shrinks

## Example

Consider the initial unsorted array `[99, 45, 35, 40, 16, 50, 11, 7, 90]`. Here's what it looks like before and after each iteration of the outer loop. The sorted sub-array is **bolded**.

> | Iteration | Array                                   |
> | --------- | --------------------------------------- |
> | 1.        | [**99**, 45, 35, 40, 16, 50, 11, 7, 90] |
> | 2.        | [**45, 99**, 35, 40, 16, 50, 11, 7, 90] |
> | 3.        | [**35, 45, 99**, 40, 16, 50, 11, 7, 90] |
> | 4.        | [**35, 40, 45, 99**, 16, 50, 11, 7, 90] |
> | 5.        | [**16, 35, 40, 45, 99**, 50, 11, 7, 90] |
> | 6.        | [**16, 35, 40, 45, 50, 99**, 11, 7, 90] |
> | 7.        | [**11, 16, 35, 40, 45, 50, 99**, 7, 90] |
> | 8.        | [**7, 11, 16, 35, 40, 45, 50, 99**, 90] |
> | 9.        | [**7, 11, 16, 35, 40, 45, 50, 90, 99**] |

Reviewer: similar changes as for previous algorithms

```suggestion
| Pass | Array                                   |
| ---- | --------------------------------------- |
| 1    | [**99**, 45, 35, 40, 16, 50, 11, 7, 90] |
| 2    | [**45, 99**, 35, 40, 16, 50, 11, 7, 90] |
| 3    | [**35, 45, 99**, 40, 16, 50, 11, 7, 90] |
| 4    | [**35, 40, 45, 99**, 16, 50, 11, 7, 90] |
| 5    | [**16, 35, 40, 45, 99**, 50, 11, 7, 90] |
| 6    | [**16, 35, 40, 45, 50, 99**, 11, 7, 90] |
| 7    | [**11, 16, 35, 40, 45, 50, 99**, 7, 90] |
| 8    | [**7, 11, 16, 35, 40, 45, 50, 99**, 90] |
| 9    | [**7, 11, 16, 35, 40, 45, 50, 90, 99**] |
```

> ## Big(O) Complexity

Reviewer: ```suggestion
## Big O Complexity
```

> In sorting, the most expensive part is a comparison of two elements. Insertion sort algorithm requires
> 
> - 0 comparisons to insert the first element
> - 1 comparison to insert the second element
> - 2 comparisons to insert the third element
> - ... and so on
> - _n-1_ comparisons (worst case) to insert the last element
> 
> Overall, this is _1 + 2 + 3 + ... + (n-1)_ = _O(n<sup>2</sup>)_
> 
> The insertion sort has a complexity of O(n<sup>2</sup>).

Reviewer: this text sounds like it's unaware that we have discussed much of this in other locations already (even in the original textbook)
add more connection to the rest of the material

```suggestion
As with the other sorting algorithms we have considered, with insertion sort we examine the number of comparisons performed in the inner loop.

The insertion sort algorithm requires (in the worst case):

- 0 comparisons to insert the first element
- 1 comparison to insert the second element
- 2 comparisons to insert the third element
- ... and so on
- _n-1_ comparisons to insert the last element

Overall, this is _1 + 2 + 3 + ... + (n-1)_, the same expression we saw for bubble sort and selection sort.

Therefore insertion sort has a complexity of _O(n<sup>2</sup>)_.
```

> Unlike selection sort, insertion sort has a best case time complexity of O(n) when the items are already sorted. In other words, insertion sort runs in nearly linear time on a nearly sorted list of elements.

Reviewer: additional explanation about why the best case is O(n)
```suggestion
We might also notice that insertion sort, like bubble sort, has a best case time complexity of _O(n)_ if the list is already sorted. As we expand the sorted sub-part of the array, each new item is already in its correct spot, which we can determine with a single check per item. Since there are _n_ items, this gives a best case of _O(n)_.

Insertion sort generally improves in complexity the closer the list is to being sorted, that is, the more items are already relatively in their correct places. In other words, insertion sort runs in _nearly_ linear time on a _nearly_ sorted list of elements.
```

> 

Reviewer: add stability information

```suggestion

## Stability

Insertion sort is a stable sorting algorithm. Like bubble sort, we only move items in the list using adjacent swaps, and then only when they are strictly out of order.

```

## Example Implementation

Consider this example implementation of insertion sort.

```python
def insertion_sort(array):
    i = 1
    while i < len(array):
        to_insert = array[i]
        j = i
        # Search in the sorted portion of the array
        # for the correct position to insert array[i]

>         while j > 0 and array[j - 1] > to_insert:
>             array[j] = array[j - 1]

Reviewer: spacing consistent with the previous algorithms
```suggestion
        while j > 0 and array[j-1] > to_insert:
            array[j] = array[j-1]
```

            j -= 1
        array[j] = to_insert
        i += 1

>     return array

Reviewer: no need for return since sorting in place

```suggestion
```

```

> - Loop through the entire list
> - Find the first element of the unsorted list. This is the element that needs to be inserted, called `to_insert`
> - Find the correct index to insert the value `to_insert`
>   - Loop through the sorted list
>   - Compare the values between the item we're inspecting, and `to_insert`
>   - Swap if needed
> - Increase the indices, so that the sorted list grows, and the unsorted list shrinks

Reviewer: include transition text used for other algorithms between example and explanation
slightly expand explanation (we're on the third one now, so a little less detail than the previous ones seems ok to me)
mention that no return is needed because in-place

```suggestion
Compare this code with this detailed explanation of the algorithm:

- Loop through the entire list with `i` representing the first position of the unsorted sub-array
- Store the first element of the unsorted list in a temporary variable called `to_insert`
- Find the correct index to insert the value `to_insert`
  - Loop through the sorted list from back to front
  - Compare the values between the item at the position we're inspecting, and `to_insert`
  - Swap if needed
- Increase the outer index `i` so that the sorted list grows, and the unsorted list shrinks

This implementation sorts the array in-place. Because the original array is modified, no return statement is needed.
```
